Conversion functions from primitive types to strings

danpoe opened the pull request:
<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/s My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

danpoe committed:
Group CProverString operations in initialize_conversion_table()

In the initialization of the conversion table, CProverString operations are
grouped together for better readability.

danpoe committed:
Add CProver.toString() functions to string preprocessing conversion table

New CProverString.toString() operations are added which are now handled
internally by cbmc. These operations can be used to model other Java methods
such as Integer.toString().

danpoe committed:
Add executable models of CProverString functions

These models can be compiled and run and have the same semantics as implemented
by the string solver. They can thus be used to check that the string solver
gives the correct result by comparing the result of a concrete run with the
result obtained by cbmc.

danpoe committed:
Add regression tests for CProverString.toString() operations

romainbrenguier commented on jbmc/src/java_bytecode/java_string_library_preprocess.cpp:
```diff
@@ -1548,6 +1548,21 @@ void java_string_library_preprocesst::initialize_conversion_table()
      "StringBuffer;II)Ljava/lang/String;"] = ID_cprover_string_substring_func;
+  cprover_equivalent_to_java_string_returning_function
+    ["java::org.cprover.CProverString.toString:(Z)Ljava/lang/String;"] =
+      ID_cprover_string_of_bool_func;
```
:pick: actually this could be implemented directly as a model, there is not complicated logic involved, so I don't know if this is needed :lollipop:

danpoe replied:
Yes, I think in that case we can remove both `ID_cprover_string_of_bool_func` as well as `CProverString.toString(char)` and use `b ? "true" : "false"` directly.

romainbrenguier commented on jbmc/src/java_bytecode/java_string_library_preprocess.cpp:
```diff
@@ -1548,6 +1548,21 @@ void java_string_library_preprocesst::initialize_conversion_table()
+      ID_cprover_string_of_bool_func;
+  cprover_equivalent_to_java_string_returning_function
+    ["java::org.cprover.CProverString.toString:(C)Ljava/lang/String;"] =
+      ID_cprover_string_of_char_func;
```
:lollipop:

danpoe replied:
We could use
```
char arr[] = {c};
return new String(arr);
```
which calls `ofCharArray()` which creates a `StringBuilder` and uses `append()` to insert the char and then returns `StringBuilder.toString()`. Any idea if this might have performance implications?

romainbrenguier commented on jbmc/src/java_bytecode/java_string_library_preprocess.cpp:
```diff
@@ -1548,6 +1548,21 @@ void java_string_library_preprocesst::initialize_conversion_table()
+      ID_cprover_string_of_long_func;
+  cprover_equivalent_to_java_string_returning_function
+    ["java::org.cprover.CProverString.toString:(F)Ljava/lang/String;"] =
+      ID_cprover_string_of_float_func;
```
why is there no string_of_double entry?

danpoe replied:
Double to string is handled by a model via `CProverString.toString(CProver.doubleToFloat(d))` as double to string conversion is not implemented AFAIK.

romainbrenguier commented on jbmc/regression/jbmc-strings/CProverStringToString/Main.java:
```diff
@@ -0,0 +1,184 @@
+    /*case 5:
+      c = Character.MAX_VALUE;
+      r = "\uFFFF";
+      break;*/
```
why is this commented out? :coffee:

danpoe replied:
This case does not work. I have to investigate why.

romainbrenguier commented on jbmc/regression/jbmc-strings/CProverStringToString/Main.java:
```diff
@@ -0,0 +1,184 @@
+    case 8:
+      f = Float.NaN;
+      r = "NaN";
+      break;*/
```
:coffee:

danpoe replied:
Does not work yet.

martin-cs commented on jbmc/regression/jbmc-strings/CProverStringToString/Main.java:
```diff
@@ -0,0 +1,184 @@
+      break;
+    case 5:
+      f = Float.MIN_VALUE;
+      r = "1.4E-45";
```
Where did this number come from?

danpoe replied:
That's the result of `String.valueOf(f)` (same as `Float.toString(f)`) :bread:

martin-cs commented on jbmc/regression/jbmc-strings/CProverStringToString/Main.java:
```diff
@@ -0,0 +1,184 @@
+      break;
+    /*case 4:
+      f = Float.MAX_VALUE;
+      r = "3.4028235E38";
```
This is kind of a weird way of formatting this.

danpoe replied:
:bread:

martin-cs reviewed (commented):
A passing thought

smowton reviewed (approved):
No further comments except what @romainbrenguier already raised

danpoe committed:
Add regression test for bug of char to string conversion

After converting '\uFFFF' to string the result differs from "\uFFFF" for a check
with `equals()`.

danpoe commented:
Ok, I'll merge this once CI passes. The behavior of the string operations does not change, but a new test has revealed some issues in the float to string conversion. I've marked the test as `KNOWNBUG` for now.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: ba122158a2b12df849dd09b1784cecfff08d8871).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/124413147
